reactor-core/src/main/java/reactor/core/publisher/FluxListen.java

```diff
@@ -0,0 +1,332 @@
+		//after it is created, in case doFirst fails we can additionally try to invoke doFinally.
+		//note that if the later handler also fails, then that exception is thrown.
+		SignalListener<T> signalListener;
+		try {
+			//TODO replace currentContext() with contextView() when available
+			signalListener = lifter.createListener(source, actual.currentContext().readOnly(), publisherState);
+		}
+		catch (Throwable generatorError) {
+			Operators.error(actual, generatorError);
+			return null;
+		}
+
+		try {
+			signalListener.doFirst();
+		}
+		catch (Throwable listenerError) {
+			signalListener.handleListenerError(listenerError);
+			Operators.error(actual, listenerError);
+			return null;
+		}
+
+		if (actual instanceof ConditionalSubscriber) {
+			//noinspection unchecked
+			return new ListenConditionalSubscriber<>((ConditionalSubscriber<? super T>) actual, signalListener);
+		}
+		return new ListenSubscriber<>(actual, signalListener);
+	}
+
+	@Nullable
+	@Override
+	public Object scanUnsafe(Attr key) {
+		if (key == Attr.RUN_STYLE) return Attr.RunStyle.SYNC;
+
+		return super.scanUnsafe(key);
+	}
+
+	//TODO support onErrorContinue around listener errors
+	static class ListenSubscriber<T> implements InnerOperator<T, T> {
+
+		final CoreSubscriber<? super T> actual;
+		final SignalListener<T>         listener;
+
+		boolean done;
+		Subscription s;
+
+		ListenSubscriber(CoreSubscriber<? super T> actual, SignalListener<T> signalListener) {
+			this.actual = actual;
+			this.listener = signalListener;
+		}
+
+		@Override
+		public CoreSubscriber<? super T> actual() {
+			return this.actual;
+		}
+
+		@Override
+		@Nullable
+		public Object scanUnsafe(Attr key) {
+			if (key == Attr.PARENT) return s;
+			if (key == Attr.TERMINATED) return done;
+			if (key == Attr.RUN_STYLE) return Attr.RunStyle.SYNC;
+
+			return InnerOperator.super.scanUnsafe(key);
+		}
+
+		/**
+		 * Cancel the prepared subscription, pass the listener error to {@link SignalListener#handleListenerError(Throwable)}
+		 * and then terminate the downstream directly with same error (without invoking any other handler).
+		 *
+		 * @param listenerError the exception thrown from a handler method before the subscription was set
+		 * @param toCancel the {@link Subscription} that was prepared but not sent downstream
+		 */
+		protected void handleListenerErrorPreSubscription(Throwable listenerError, Subscription toCancel) {
+			toCancel.cancel();
+			listener.handleListenerError(listenerError);
+			Operators.error(actual, listenerError);
+		}
+
+		/**
+		 * Cancel the active subscription, pass the listener error to {@link SignalListener#handleListenerError(Throwable)}
+		 * and then terminate the downstream directly with same error (without invoking any other handler).
+		 *
+		 * @param listenerError the exception thrown from a handler method
+		 */
+		protected void handleListenerErrorAndTerminate(Throwable listenerError) {
```

Comment:
do we still support continuation? Probably we have to check if we wanna continue here before cancelling subscription

Reply:
we might want to add onErrorContinue support to the operator in some cases, but that's not an absolute priority for M2